applications/tari_indexer/src/event_scanner.rs

```diff
@@ -0,0 +1,561 @@
+        models::{
+            events::{NewEvent, NewScannedBlockId},
+            substate::NewSubstate,
+        },
+        sqlite_substate_store_factory::{
+            SqliteSubstateStore,
+            SubstateStore,
+            SubstateStoreReadTransaction,
+            SubstateStoreWriteTransaction,
+        },
+    },
+};
+
+const LOG_TARGET: &str = "tari::indexer::event_scanner";
+
+#[derive(Default, Debug, Clone)]
+pub struct EventFilter {
+    pub topic: Option<String>,
+    pub entity_id: Option<EntityId>,
+    pub substate_id: Option<SubstateId>,
+    pub template_address: Option<TemplateAddress>,
+}
+
+impl TryFrom<EventFilterConfig> for EventFilter {
+    type Error = anyhow::Error;
+
+    fn try_from(cfg: EventFilterConfig) -> Result<Self, Self::Error> {
+        let entity_id = cfg.entity_id.map(|str| EntityId::from_hex(&str)).transpose()?;
+        let substate_id = cfg.substate_id.map(|str| SubstateId::from_str(&str)).transpose()?;
+        let template_address = cfg
+            .template_address
+            .map(|str| TemplateAddress::from_str(&str))
+            .transpose()?;
+
+        Ok(Self {
+            topic: cfg.topic,
+            entity_id,
+            substate_id,
+            template_address,
+        })
+    }
+}
+
+pub struct EventScanner {
+    network: Network,
+    epoch_manager: Box<dyn EpochManagerReader<Addr = PeerAddress>>,
+    client_factory: TariValidatorNodeRpcClientFactory,
+    substate_store: SqliteSubstateStore,
+    event_filters: Vec<EventFilter>,
+}
+
+impl EventScanner {
+    pub fn new(
+        network: Network,
+        epoch_manager: Box<dyn EpochManagerReader<Addr = PeerAddress>>,
+        client_factory: TariValidatorNodeRpcClientFactory,
+        substate_store: SqliteSubstateStore,
+        event_filters: Vec<EventFilter>,
+    ) -> Self {
+        Self {
+            network,
+            epoch_manager,
+            client_factory,
+            substate_store,
+            event_filters,
+        }
+    }
+
+    pub async fn scan_events(&self) -> Result<usize, anyhow::Error> {
+        info!(
+            target: LOG_TARGET,
+            "scan_events",
+        );
+
+        let mut event_count = 0;
+
+        let current_epoch = self.epoch_manager.current_epoch().await?;
+        // let network_committee_info = self.epoch_manager.get_network_committees().await?;
+        // let epoch = network_committee_info.epoch;
+        let current_committees = self.epoch_manager.get_committees(current_epoch).await?;
+        for (shard, committee) in current_committees {
+            info!(
+                target: LOG_TARGET,
+                "Scanning committee epoch={}, shard={}",
+                current_epoch,
+                shard
+            );
+            // TODO: use the latest block id that we scanned for each committee
+            let new_blocks = self
+                .get_new_blocks_from_committee(shard, &mut committee.clone(), current_epoch)
```

Comment:
Unnecessary clone - mark `committee` as mutable above
```suggestion
                .get_new_blocks_from_committee(shard, &mut committee, current_epoch)
```